### PR TITLE
Use tokenize in dask.bag

### LINKED ISSUE #950: dask.bag does not use hashed keys
We should use `tokenize` rather than the current `tokens` within dask.bag

cc @jcrist

### PR DESCRIPTION
**Pull request by jcrist**
Use tokenize in dask.bag to ensure consistent names. Note that I currently use `uuid`s for some things that seemed tricky to ensure were consistent across multiple runs, mainly dealing with files/remote sources (no guarantee these haven't changed between executions).

Fixes #950.

### CODE REVIEW

**Comment by mrocklin** [comment]
In dask.dataframe.io we tokenize by the filename and the os provided modification time.  This seems to have worked well so far

**Comment by mrocklin** [comment]
We can nuke the from_hdfs function

**Comment by jcrist** [comment]
I know we use the os time. I was mainly concerned with doing that, in the case the filenames aren't on the box creating the bag. Dataframe requires the filename be local to sniff out the dtypes, bag doesn't open anything until runtime.

File: dask/bag/core.py [hunk 1]
```diff
@@ -482,19 +481,22 @@ def reduction(self, perpartition, aggregate, split_every=None,
         """
         if split_every is None:
             split_every = 8
-        a = next(names)
-        b = next(names)
+        token = tokenize(self, perpartition, aggregate, split_every)
+        a = 'reduction-part-' + token
```
Comment (mrocklin) [inline dask/bag/core.py]: For user diagnostics it would be good to use a more informative name here, perhaps `dask.utils.funcname(perpartition)`

Comment (jcrist) [inline dask/bag/core.py]: I find I look at the tasks more than the task names in the diagnostics, which show the function name. Happy to change though.

**Comment by mrocklin** [comment]
It's a valid concern, and I guess it makes sense to be cautious by default.

File: dask/bag/core.py (hunk 1 above)
Comment (mrocklin) [inline dask/bag/core.py]: In `distributed` the client only has access to the keyname, not the full task.

Comment (jcrist) [inline dask/bag/core.py]: Good to know. I updated a few other ones that could be more verbose.

**Comment by mrocklin** [comment]
+1

**Pull request merged**